scripts/checkpoint_averaging/torch_distributed_checkpoint_averaging.py

```diff
@@ -0,0 +1,253 @@
+                if key in ckpt_dir:
+                    checkpoint_paths.append(ckpt_dir)
+
+    checkpoint_paths = [p for p in checkpoint_paths if not p.endswith(".nemo")]
+    n = len(checkpoint_paths)
+
+    logging.info(f"Averaging {n} checkpoints ... {'at steps:' + str(args.steps) if args.steps is not None else ''}")
+
+    avg_weights = {}
+    with torch.no_grad():
+        for ix, path in tqdm(enumerate(checkpoint_paths), total=len(checkpoint_paths)):
+            full_path = os.path.join(args.checkpoint_dir, path)
+            weights, metadata = load_distributed_weights_dcp(full_path)
+            for k, v in weights.items():
+                if "_extra_state" in k:  # Extra state is not averaged
+                    if ix == 0:
+                        avg_weights[k] = v
+                    continue
+                if k.startswith(
+                    "optimizer."
+                ):  # These should be filtered out during load but skipping them here just for safety
+                    continue
+                if k not in avg_weights:
+                    logging.info(f'"Initialized average weights dict with: {k}"')
+                    avg_weights[k] = v.data.clone().detach()
+                else:
+                    avg_weights[k] += v.data.clone().detach()
+            del weights
+            gc.collect()
+
+        for k in avg_weights:
+            if "_extra_state" in k:
+                continue
+            if str(avg_weights[k].dtype).startswith("int"):
+                raise ValueError("Int type not supported")
+            else:
+                array = avg_weights[k] / n
+                avg_weights[k] = array
+
+        logging.info(f"Finished averaging {n} checkpoints")
+
+    if args.steps is None:
+        ckpt_name = os.path.join(args.checkpoint_dir, args.name_prefix + "-averaged", "model_weights")
+    else:
+        steps_combined = "_".join([str(x) for x in args.steps])
+        ckpt_name = os.path.join(
+            args.checkpoint_dir,
+            args.name_prefix + "-" + steps_combined + "-averaged",
+            "model_weights",
+        )
+
+    for k, v in metadata.items():
+        if isinstance(v, ShardedTensor):
+            averaged_weight = avg_weights[k]
+            v.data = averaged_weight
+            avg_weights[k] = v
+        elif isinstance(v, ShardedObject):  # BytesIO, not averaged
+            data = avg_weights[k]
+            v.data = data
+            avg_weights[k] = v
+
+    save_distributed_weights_dcp(ckpt_name, avg_weights)
+    if dist.get_rank() == 0:
+        ckpt_name = os.path.dirname(ckpt_name)
+        shutil.copy(
+            os.path.join(args.untarred_nemo_folder, "model_config.yaml"),
+            os.path.join(ckpt_name, "model_config.yaml"),
+        )
+
+        files = glob.glob(f"{args.untarred_nemo_folder}/*.model") + glob.glob(
+            f"{args.untarred_nemo_folder}/*.vocab.json"
+        )
+        logging.info(f"Copying other files: {files}")
```

Comment:
Do we want to copy the original `common.pt` (which contains all the PTL training related data like params, loops state etc.)?

Reply:
If the averaged ckpts are used for inference/fine-tuning then probably not :+1: